packages/wallet/src/redux/protocols/existing-channel-funding/reducer.ts

```
@@ -0,0 +1,349 @@
+  proposedAmount: string,
+  channelId: string,
+): { proposedAllocation: string[]; proposedDestination: string[] } {
+  const numParticipants = latestCommitment.channel.participants.length;
+  const amountRequiredFromEachParticipant = bigNumberify(proposedAmount).div(numParticipants);
+
+  const proposedAllocation: string[] = [];
+  const proposedDestination: string[] = [];
+
+  for (let i = 0; i < latestCommitment.allocation.length; i++) {
+    const allocation = latestCommitment.allocation[i];
+
+    const newAmount = bigNumberify(allocation).sub(amountRequiredFromEachParticipant);
+
+    if (newAmount.gt('0x0')) {
+      proposedAllocation.push(newAmount.toHexString());
+      proposedDestination.push(latestCommitment.destination[i]);
+    }
+  }
+
+  proposedAllocation.push(proposedAmount);
+  proposedDestination.push(channelId);
+
+  return { proposedAllocation, proposedDestination };
+}
+
+function ledgerChannelNeedsTopUp(latestCommitment: Commitment, proposedAmount: string) {
+  if (latestCommitment.commitmentType !== CommitmentType.App) {
+    throw new Error('Ledger channel is already closed.');
+  }
+  const numParticipants = latestCommitment.channel.participants.length;
+  const amountRequiredFromEachParticipant = bigNumberify(proposedAmount).div(numParticipants);
+
+  return !latestCommitment.allocation.every(a =>
+    bigNumberify(a).gte(amountRequiredFromEachParticipant),
+  );
+}
+
+function craftAndSendAppPostFundCommitment(
+  sharedData: SharedData,
+  appChannelId: string,
+  processId: string,
+): SharedData | 'CouldNotSign' | 'NotASetupCommitment' {
+  let newSharedData = { ...sharedData };
+  const appChannel = getChannel(sharedData, appChannelId);
+  if (!appChannel) {
+    throw new Error(`Could not find application channel ${appChannelId}`);
+  }
+
+  const theirAppCommitment = getLastCommitment(appChannel);
+
+  const ourAppCommitment = nextSetupCommitment(theirAppCommitment);
+  if (ourAppCommitment === 'NotASetupCommitment') {
```

Perhaps `nextSetupCommitment` should throw instead of return a string, and always return a `Commitment`

---

I was debating that as well. That probably makes more sense in hindsight.